variable.c

```diff
@@ -1463,6 +1463,36 @@ rb_ivar_get(VALUE obj, ID id)
     return iv;
 }
 
+VALUE
+rb_ivar_get_at(VALUE obj, attr_index_t index, ID id)
+{
+    RUBY_ASSERT(rb_is_instance_id(id));
+    // Used by JITs, but never for T_OBJECT.
+
+    switch (BUILTIN_TYPE(obj)) {
+      case T_OBJECT:
+        UNREACHABLE_RETURN(Qundef);
+      case T_CLASS:
+      case T_MODULE:
+        {
+            VALUE fields_obj = RCLASS_WRITABLE_FIELDS_OBJ(obj);
+            VALUE val = rb_imemo_fields_ptr(fields_obj)[index];
+
+            if (UNLIKELY(!rb_ractor_main_p()) && !rb_ractor_shareable_p(val)) {
+                rb_raise(rb_eRactorIsolationError,
+                        "can not get unshareable values from instance variables of classes/modules from non-main Ractors");
+            }
+
+            return val;
+        }
```

Comment:
It's probably nice to put the unreachable case later so that `T_CLASS` and `T_MODULE` cases can run without checking if it's `T_OBJECT`.

```suggestion
      case T_CLASS:
      case T_MODULE:
        {
            VALUE fields_obj = RCLASS_WRITABLE_FIELDS_OBJ(obj);
            VALUE val = rb_imemo_fields_ptr(fields_obj)[index];

            if (UNLIKELY(!rb_ractor_main_p()) && !rb_ractor_shareable_p(val)) {
                rb_raise(rb_eRactorIsolationError,
                        "can not get unshareable values from instance variables of classes/modules from non-main Ractors");
            }

            return val;
        }
      case T_OBJECT:
        UNREACHABLE_RETURN(Qundef);
```

Reply:
From the name, I assumed there are cases where it's compiled to `return`. I see that it's probably `__builtin_unreachable()` for environments we primarily care about, but we also seem to support environments that don't have one.

Reply:
Yeah, the unreacheable basically eliminate the branch.